ownCloud/Settings/Passcode/PasscodeManager.swift

```diff
@@ -0,0 +1,375 @@
+        case addPasscodeSecondStep
+        case unlockPasscode
+        case unlockPasscodeError
+        case deletePasscode
+        case deletePasscodeError
+        case addPasscodeFirstStepAfterErrorOnSecond
+    }
+
+    // MARK: Global vars
+
+    // Common
+    private var window: LockWindow?
+    private let passcodeLength = 4
+    private let passcodeKeychainAccount = "passcode-keychain-account"
+    private let passcodeKeychainPath = "passcode-keychain-path"
+    private var passcodeMode: PasscodeInterfaceMode?
+    private var passcodeViewController: PasscodeViewController?
+
+    // Add/Delete
+    private var passcodeFromFirstStep: String?
+    private var completionHandler: CompletionHandler?
+
+    // Unlock
+    private var datePressedHomeButton: Date?
+    private var userDefaults: UserDefaults?
+
+    // Brute force protection
+    public let TimesPasscodeFailedKey: String =  "times-passcode-failed"
+    public let DateAllowTryPasscodeAgainKey: String =  "date-allow-try-passcode-again"
+    private var timesPasscodeFailed: Int {
+        didSet {
+            self.userDefaults!.set(timesPasscodeFailed, forKey: TimesPasscodeFailedKey)
+        }
+    }
+    private var dateAllowTryAgain: Date? {
+        didSet {
+            self.userDefaults!.set(NSKeyedArchiver.archivedData(withRootObject: dateAllowTryAgain as Any), forKey: DateAllowTryPasscodeAgainKey)
+        }
+    }
+    private let timesAllowPasscodeFail: Int = 3
+    private let multiplierBruteForce: Int = 10
+    private var timerBruteForce: Timer?
+
+    // Utils
+    var isPasscodeActivated: Bool {
+        return (self.userDefaults!.bool(forKey: SecuritySettingsPasscodeKey) && isPasscodeStoredOnKeychain)
+    }
+
+    var isPasscodeStoredOnKeychain: Bool {
+        return (OCAppIdentity.shared().keychain.readDataFromKeychainItem(forAccount: passcodeKeychainAccount, path: passcodeKeychainPath) != nil)
+    }
+
+    private var passcodeFromKeychain: String? {
+        if let passcodeData = OCAppIdentity.shared().keychain.readDataFromKeychainItem(
+            forAccount: passcodeKeychainAccount, path: passcodeKeychainPath) {
+            return NSKeyedUnarchiver.unarchiveObject(with: passcodeData) as? String
+        } else {
+            return nil
+        }
+    }
+
+    private var shouldBeLocked: Bool {
+        var output: Bool = true
+
+        if isPasscodeActivated {
+            if let date = self.datePressedHomeButton {
+
+                let elapsedSeconds = Date().timeIntervalSince(date)
+                let minSecondsToAsk = self.userDefaults?.integer(forKey: SecuritySettingsFrequencyKey)
+
+                if Int(elapsedSeconds) < minSecondsToAsk! {
+                    output = false
+                }
+            }
+        } else {
+            output = false
+        }
+
+        return output
+    }
+
+    // MARK: - Init
+
+    static var sharedPasscodeManager = PasscodeManager()
+
+    public override init() {
```

Comment:
The UserDefaults it is seted just once on the init of the singleton. If in every call to the singleton I have to pass the same parameter will be ugly.

The UserDefaults should be centralice just in one point to be read (we talked about it before that it will be on the SDK) and it is on that point where it should be mocked.

Reply:
IMO doesn't matter if we're going to use the SDK's userDefaults or another one. It's easier for people like @jesmrec to mock a single object here than mock the SDK's one.

Of course, there should be other parts of the app that we should mock subclassing the SDK, but if we can avoid hardcoding here the user default and passing it by parameter could help us to make clean states easier for tests.

Taking a look at why this is a singleton, you are relying on the keychain and the user defaults, so you don't need a singleton with a global state, you can create an object every time you need it then pass the userDefaults you want as an argument and finally configure your view with the current state of the userDefaults and the keychain.